### PR TITLE
Added package for verifying reports on non-TEE environments.

### PR DESCRIPTION
**Pull request by gupta-ak**
This is for Linux only atm. Once the windows attestation PRs are in, we can add Windows support.

To build the package, all you simply do the same process as building the regular OE SDK package aside from the last CPack step, which becomes:
```
cpack -D CPACK_DEB_COMPONENT_INSTALL=ON -DCPACK_COMPONENTS_ALL=OEHOSTVERIFY
```
The CPack command for building the OE SDK remains the same.

Some pending open issues:

1. Name of the package. We internally refer it as the "host verification library" but that probably doesn't make sense to external people.
2. Expected behavior if both the OE SDK and host verification library is installed.

### CODE REVIEW

File: host/CMakeLists.txt [hunk 1]
```diff
@@ -253,6 +283,7 @@ endif ()
 
 # For including edge routines.
 target_include_directories(oehost PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
+target_include_directories(oehostverify PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
```
Comment (CodeMonkeyLeet) [inline host/CMakeLists.txt]: This shouldn't be necessary since there's no dependency on oeedger8r generated headers, no?

File: host/CMakeLists.txt [hunk 2]
```diff
@@ -337,8 +392,24 @@ if (UNIX)
       -Wl,-z,noexecstack)
 endif ()
 
+add_library(oehostverifyapp INTERFACE)
```
Comment (CodeMonkeyLeet) [inline host/CMakeLists.txt]: Do we need two versions of the `oehostverify` library given the use case? Should the dynamic export version just be the default version if that's needed? I don't believe we publish `oehostapp` in the SDK today, and it's consumed by tests only.

File: host/sgx/report.c [hunk 3]
```diff
@@ -293,8 +294,7 @@ oe_result_t oe_verify_report(
         OE_CHECK(oe_initialize_quote_provider());
 
         // Quote attestation can be done entirely on the host side.
-        OE_CHECK(VerifyQuoteImpl(
-            header->report, header->report_size, NULL, 0, NULL, 0, NULL, 0));
+        OE_CHECK(oe_verify_remote_report(report, report_size, parsed_report));
```
[the remark this answers is not in the document]
Comment (gupta-ak) [reply host/sgx/report.c]: I can revert it back to `VerifyQuoteImpl` since there's a bunch of overhead. I'll rename that function too so that it matches the OE conventions.

File: host/CMakeLists.txt (hunk 2 above)
Comment (gupta-ak) [reply host/CMakeLists.txt]: The dynamic library could be the default since I don't see the point of the static one. The `oehostapp` is part of the export list, so I think it's a usable target by the external consumers.

File: host/CMakeLists.txt [hunk 4]
```diff
@@ -95,16 +98,22 @@ else()
...
     ../common/sgx/quote.c
     ../common/sgx/report.c
     ../common/sgx/revocation.c
     ../common/sgx/sgxcertextensions.c
     ../common/sgx/tcbinfo.c
-    ../common/sgx/tlsverifier.c
+    sgx/sgxquoteprovider.c
+    sgx/hostverify_report.c)
+
+  list(APPEND PLATFORM_HOSTVERIFY_SRC ${NEW_HOSTVERIFY_SRC})
```
[the remark this answers is not in the document]
Comment (gupta-ak) [reply host/CMakeLists.txt]: Agreed. I'll try to simply it. Ideally, we should simplify it even further; the platform specific folders should have their own CMakeLists.txt so that we can reduce the number of branches in the top level host CMake.

File: host/CMakeLists.txt (hunk 1 above)
Comment (gupta-ak) [reply host/CMakeLists.txt]: Good point. It shouldn't be necessary.

File: host/sgx/report.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (gupta-ak) [reply host/sgx/report.c]: Done

File: cmake/cpack_settings.cmake [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+# CPack variables for the regular OE SDK.
+include(InstallRequiredSystemLibraries)
+set(CPACK_PACKAGE_NAME "open-enclave")
+set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Open Enclave SDK")
+set(CPACK_PACKAGE_CONTACT "openenclave@microsoft.com")
+set(CPACK_PACKAGE_DESCRIPTION_FILE "${PROJECT_SOURCE_DIR}/README.md")
+set(CPACK_RESOURCE_FILE_LICENSE "${PROJECT_SOURCE_DIR}/LICENSE")
+set(CPACK_PACKAGE_VERSION ${OE_VERSION})
+set(CPACK_PACKAGING_INSTALL_PREFIX ${CMAKE_INSTALL_PREFIX})
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "libsgx-enclave-common (>=2.3.100.46354-1), libsgx-enclave-common-dev (>=2.3.100.0-1), libsgx-dcap-ql (>=1.0.100.46460-1.0), libsgx-dcap-ql-dev (>=1.0.100.46460-1.0), pkg-config")
```
Comment (andschwa) [inline cmake/cpack_settings.cmake]: Just curious, why do these _depend_ on pkg-config instead of just recommending it? Do they need it at runtime to find other packages?

File: cmake/package_settings.cmake [hunk 6]
```diff
@@ -39,30 +39,21 @@ write_basic_package_version_file(
...
+  COMPONENT OEHOSTVERIFY)
 install(
   EXPORT openenclave-targets
   NAMESPACE openenclave::
   # Note that this is used in `openenclaverc` to set the path for
   # users of the SDK and so must remain consistent.
   DESTINATION ${CMAKE_INSTALL_LIBDIR}/openenclave/cmake
-  FILE openenclave-targets.cmake)
+  FILE openenclave-targets.cmake
+  COMPONENT OEHOSTVERIFY)
```
Comment (andschwa) [inline cmake/package_settings.cmake]: What scenario _won't_ these be required? (I'm trying to understand the choice of components.)

File: host/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (andschwa) [reply host/CMakeLists.txt]: I like that approach, although I'd suggest that you could do it separately from this PR, as it's a bit of technical debt cleanup.

File: host/CMakeLists.txt [hunk 7]
```diff
@@ -300,6 +331,17 @@ if (USE_LIBSGX)
 endif ()
 
 # Compile definitions and options
+target_compile_definitions(oehostverify
+  PUBLIC
+  # NOTE: This definition is public to the rest of our project's
+  # targets, but should not yet be exposed to consumers of our
+  # package.
+  $<BUILD_INTERFACE:OE_API_VERSION=2>
```
Comment (andschwa) [inline host/CMakeLists.txt]: Just a random question: when _are_ we going to make this public? This release?

File: host/CMakeLists.txt [hunk 8]
```diff
@@ -325,13 +375,18 @@ if (UNIX)
...
 if (CMAKE_C_COMPILER_ID MATCHES GNU)
+  target_compile_options(oehostverify PRIVATE -Wjump-misses-init)
   target_compile_options(oehost PRIVATE -Wjump-misses-init)
 endif ()
 
 # TODO: Remove this hard coded output directory.
 set_property(TARGET oehost PROPERTY
   ARCHIVE_OUTPUT_DIRECTORY ${OE_LIBDIR}/openenclave/host)
 
+# TODO: Remove this hard coded output directory.
```
Comment (andschwa) [inline host/CMakeLists.txt]: Why did you add it? 😂 I don't believe it's necessary to be bin-placing these things.

File: cmake/cpack_settings.cmake (hunk 5 above)
Comment (gupta-ak) [reply cmake/cpack_settings.cmake]: Good point. We shouldn't need to depend on pkgconfig (i.e. it's not necessary to use the SDK).

File: cmake/package_settings.cmake (hunk 6 above)
Comment (gupta-ak) [reply cmake/package_settings.cmake]: It will always be required I think. To create the full SDK, you do a non-component CPack, which will include everything. To create the host verification library, you do a component install w/ component=OEHOSTVERIFY.

File: host/CMakeLists.txt (hunk 7 above)
Comment (gupta-ak) [reply host/CMakeLists.txt]: no idea.

File: host/CMakeLists.txt (hunk 8 above)
Comment (gupta-ak) [reply host/CMakeLists.txt]: cuz I copied and pasted a bunch of it 😂

File: host/CMakeLists.txt (hunk 7 above)
Comment (andschwa) [reply host/CMakeLists.txt]: 🤷‍♂️

File: host/CMakeLists.txt (hunk 2 above)
Comment (gupta-ak) [reply host/CMakeLists.txt]: Made the dynamic export the default. That's actually how it is in our pkgconfig too.

File: host/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (gupta-ak) [reply host/CMakeLists.txt]: I simplified the cmake a bit to make it easier to follow, but we should fix it properly later. I'll make an issue for it.

File: host/CMakeLists.txt (hunk 7 above)
Comment (CodeMonkeyLeet) [reply host/CMakeLists.txt]: @radhikaj we should track the answer to this question.

File: docs/GettingStartedDocs/Contributors/InstallInfo.md [hunk 9]
```diff
@@ -64,3 +64,14 @@ as above. For example, to create a Debian package that will install the SDK to
 cmake -DCMAKE_INSTALL_PREFIX=/opt/openenclave ..
 cpack -G DEB
 ```
+
+## Create the Host-only Report Verification package
```
Comment (CodeMonkeyLeet) [inline docs/GettingStartedDocs/Contributors/InstallInfo.md]: Copyedit: Is "Host-only Report Verification" package a proper noun? I would follow the existing casing convention and go with:
> ## Create the host-only report verification package

File: docs/GettingStartedDocs/Contributors/InstallInfo.md [hunk 10]
```diff
@@ -64,3 +64,14 @@ as above. For example, to create a Debian package that will install the SDK to
...
+
+The host report verification package allows non-enclave applications to
```
Comment (CodeMonkeyLeet) [inline docs/GettingStartedDocs/Contributors/InstallInfo.md]: Copyedit: Consider consistent terminology, so "host-only report verification package" here as well

File: docs/GettingStartedDocs/Contributors/InstallInfo.md [hunk 11]
```diff
@@ -64,3 +64,14 @@ as above. For example, to create a Debian package that will install the SDK to
...
+validate remote reports from enclaves. The process to create this package
+is almost the same as the normal OE SDK. The only difference is the `cpack`
+command as shown below:
+
+```bash
+cpack -G DEB -D CPACK_DEB_COMPONENT_INSTALL=ON  -D CPACK_COMPONENTS_ALL=OEHOSTVERIFY
```
Comment (CodeMonkeyLeet) [inline docs/GettingStartedDocs/Contributors/InstallInfo.md]: Nit: extra space at `=ON  -D CPACK_`

File: cmake/cpack_settings.cmake [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+# CPack variables for the non-enclave host verification package.
+set(CPACK_DEBIAN_OEHOSTVERIFY_PACKAGE_NAME "open-enclave-hostverify")
+set(CPACK_DEBIAN_OEHOSTVERIFY_FILE_NAME DEB-DEFAULT)
```
[the remark this answers is not in the document]
Comment (CodeMonkeyLeet) [reply cmake/cpack_settings.cmake]: Can we file an issue for this if not addressed?

[the remark this answers is not in the document]
Comment (gupta-ak) [reply cmake/cpack_settings.cmake]: oops, missed this. I made it match the OE SDK's naming convention.

File: docs/GettingStartedDocs/Contributors/InstallInfo.md (hunk 9 above)
Comment (gupta-ak) [reply docs/GettingStartedDocs/Contributors/InstallInfo.md]: Fixed

File: docs/GettingStartedDocs/Contributors/InstallInfo.md (hunk 10 above)
Comment (gupta-ak) [reply docs/GettingStartedDocs/Contributors/InstallInfo.md]: Fixed

File: docs/GettingStartedDocs/Contributors/InstallInfo.md (hunk 11 above)
Comment (gupta-ak) [reply docs/GettingStartedDocs/Contributors/InstallInfo.md]: Fixed

File: cmake/cpack_settings.cmake (hunk 12 above)
[the remark this answers is not in the document]
Comment (gupta-ak) [reply cmake/cpack_settings.cmake]: I created https://github.com/openenclave/openenclave/issues/2107 for it.

File: cmake/cpack_settings.cmake (hunk 5 above)
Comment (gupta-ak) [reply cmake/cpack_settings.cmake]: I created https://github.com/openenclave/openenclave/issues/2107 for it

File: host/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (gupta-ak) [reply host/CMakeLists.txt]: Made the issue here: https://github.com/openenclave/openenclave/issues/2108

File: host/CMakeLists.txt [hunk 13]
```diff
@@ -95,16 +98,22 @@ else()
   message(FATAL_ERROR "Unknown OS. Only supported OSes are Linux and Windows")
 endif()
 
-# SGX specific files
+# SGX specific files.
 if (OE_SGX)
-  list(APPEND PLATFORM_SRC
+  set(NEW_HOSTVERIFY_SRC
```
[the remark this answers is not in the document]
Comment (gupta-ak) [reply host/CMakeLists.txt]: i think this is superseded by the "bottom-up" approach mentioned in the issue here: https://github.com/openenclave/openenclave/issues/2108

**Comment by gupta-ak** [comment]
@CodeMonkeyLeet Here is the open issue list that we should finish between we officially release: https://github.com/openenclave/openenclave/issues/2109
